gnomad_qc/v3/create_release/prepare_vcf_data_release.py

```diff
@@ -0,0 +1,1019 @@
+    :param vcf_info_reorder: Optional list of INFO fields to reorder, the rest of the fields are added after this list.
+    :return: Prepared HT for sanity checks and VCF export
+    """
+    logger.info("Starting preparation of VCF HT...")
+    logger.info("Adding non-PAR annotation...")
+    region_flag_expr = ht.region_flag.annotate(
+        nonpar=(ht.locus.in_x_nonpar() | ht.locus.in_y_nonpar())
+    )
+
+    logger.info(
+        "Unfurling nested gnomAD frequency annotations and add to INFO field..."
+    )
+    info_struct, freq_entries_to_remove = unfurl_nested_annotations(
+        ht,
+        entries_to_remove=freq_entries_to_remove,
+        is_subset=is_subset,
+        gnomad_release=False,
+    )
+
+    if is_subset:
+        logger.info("Adding gnomAD frequency information...")
+        gnomad_info_struct, freq_entries_to_remove = unfurl_nested_annotations(
+            ht,
+            entries_to_remove=freq_entries_to_remove,
+            is_subset=is_subset,
+            gnomad_release=True,
+        )
+        info_struct = info_struct.annotate(**gnomad_info_struct)
+
+    # NOTE: Merging rsid set into a semi-colon delimited string
+    # dbsnp might have multiple identifiers for one variant
+    # thus, rsid is a set annotation, starting with version b154 for dbsnp resource:
+    # https://github.com/broadinstitute/gnomad_methods/blob/master/gnomad/resources/grch38/reference_data.py#L136
+    # `export_vcf` expects this field to be a string, and vcf specs
+    # say this field may be delimited by a semi-colon:
+    # https://samtools.github.io/hts-specs/VCFv4.2.pdf
+    logger.info("Reformatting rsid...")
+    rsid_expr = hl.str(";").join(ht.rsid)
+
+    logger.info("Reformatting VEP annotation...")
+    vep_expr = vep_struct_to_csq(ht.vep)
+
+    # Add variant annotations to INFO field
+    # This adds annotations from:
+    #   RF struct, VQSR struct, allele_info struct,
+    #   info struct (site and allele-specific annotations),
+    #   region_flag struct, and
+    #   raw_qual_hists/qual_hists structs.
+    logger.info("Constructing INFO field")
+    ht = ht.annotate(
+        region_flag=region_flag_expr,
+        release_ht_info=ht.info,
+        info=info_struct,
+        rsid=rsid_expr,
+        vep=vep_expr,
+    )
+
+    if is_subset:
+        info_expr = make_info_expr(ht, hist_prefix="gnomad")
+    else:
+        info_expr = make_info_expr(ht)
+
+    ht = ht.annotate(info=ht.info.annotate(**info_expr, vep=ht.vep))
+
+    if freq_entries_to_remove:
+        ht = ht.annotate_globals(
+            vep_csq_header=VEP_CSQ_HEADER, freq_entries_to_remove=freq_entries_to_remove
+        )
+    else:
+        ht = ht.annotate_globals(
+            vep_csq_header=VEP_CSQ_HEADER,
+            freq_entries_to_remove=hl.empty_array(hl.tstr),
+        )
+
+    # Select relevant fields for VCF export
+    ht = ht.select("info", "filters", "rsid")
```

Comment:
Yes, we do not have a qual annotation. I think for UKBB we ended up using qualapprox right? I can add a note, but I would like to keep as is since v3 and v3.1 didn't have this